src/Kernel-Tests/SpecialSelectorTest.class.st

```diff
@@ -0,0 +1,119 @@
+"
+Tests related to special selector used bu the VM for some callback messages
+"
+Class {
+	#name : #SpecialSelectorTest,
+	#superclass : #TestCase,
+	#category : #'Kernel-Tests-Special'
+}
+
+{ #category : #tests }
+SpecialSelectorTest >> testAttemptToAssignWithIndex [
+
+	| object |
+	object := TestSpecialSelector new.
+	"Set the ivar before we are forbideen to assign.
+	We will mutate it instead of (attempt) to assign it."
+	object lastHook: nil->nil.
+
+	object value: 1.
+	self assert: object value equals: 1.
+	self assert: object lastHook equals: nil->nil.
+	
+	object setIsReadOnlyObject: true.
+	object value: 20.
+	self assert: object value equals: 1.
+	self assert: object lastHook key equals: #attemptToAssign:withIndex:.
+	self assert: object lastHook value first equals: 20.
+	self assert: object lastHook value second equals: 1. "ivar index of value"
+
+	object setIsReadOnlyObject: false.
+	object value: 3.
+	self assert: object value equals: 3.
+]
+
+{ #category : #tests }
+SpecialSelectorTest >> testCannotInterpret [
+	"Unfortunately, cannotInterpret: is currently broken.
+	The result of `cannotInterpret:` seems to be used as a receiver for the original message AGAIN.
+	This is likely a VM bug :("
+
+	| class object result |
+	"Create a new anonymous subclass of TestSpecialSelector with a nil methodDict"
+	class := Behavior new.
+	class basicSuperclass: TestSpecialSelector.
+	class methodDict: nil.
+	object := class new.
+	"Any message send to `object` should callback cannotInterpret:"
+	result := object max: 42.
+	self assert: result key equals: #cannotInterpret:.
+	self assert: result value selector equals: #max:.
+	self assert: result value arguments equals: #( 42 ).
+	self assert: result value lookupClass equals: TestSpecialSelector
+]
+
+{ #category : #tests }
+SpecialSelectorTest >> testCannotInterpretProtoObject [
+	"Test the implementation of ProtoObject>>cannotInterpret:
+	Is currenlty broken and goes in an infinite recursion of Context>>nextHandlerContext"
+
+	| class object |
+	"Create a new anonymous subclass of ProtoObject with a nil methodDict"
+	class := Behavior new.
+	class basicSuperclass: ProtoObject.
+	class methodDict: nil.
+	object := class new.
```

Comment:
I got a bug:
* here I call `new`
* that will call `initialize` on the instance
* that cannot be looked up by the VM because `methDict` is nil
* that will call `cannotInterpret:`
* that will do a manual image-side lookup and find `Object>>initialize`
* that will call `Message>>SentTo:`
* that will call `perform:withArguments:inSuperclass:` on the original receiver
* that cannot be looked up by the VM because `methDict` is nil
* while true...

solution: use `basicNew`